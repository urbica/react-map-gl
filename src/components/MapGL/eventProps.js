// @flow

export type EventProps = {
  /** The resize event handler */
  onResize?: Function,

  /** The remove event handler */
  onRemove?: Function,

  /** The mousedown event handler */
  onMousedown?: Function,

  /** The mouseup event handler */
  onMouseup?: Function,

  /** The mouseover event handler */
  onMouseover?: Function,

  /** The mousemove event handler */
  onMousemove?: Function,

  /** The click event handler */
  onClick?: Function,

  /** The dblclick event handler */
  onDblclick?: Function,

  /** The mouseenter event handler */
  onMouseenter?: Function,

  /** The mouseleave event handler */
  onMouseleave?: Function,

  /** The mouseout event handler */
  onMouseout?: Function,

  /** The contextmenu event handler */
  onContextmenu?: Function,

  /** The touchstart event handler */
  onTouchstart?: Function,

  /** The touchend event handler */
  onTouchend?: Function,

  /** The touchmove event handler */
  onTouchmove?: Function,

  /** The touchcancel event handler */
  onTouchcancel?: Function,

  /** The movestart event handler */
  onMovestart?: Function,

  /** The move event handler */
  onMove?: Function,

  /** The moveend event handler */
  onMoveend?: Function,

  /** The dragstart event handler */
  onDragstart?: Function,

  /** The drag event handler */
  onDrag?: Function,

  /** The dragend event handler */
  onDragend?: Function,

  /** The zoomstart event handler */
  onZoomstart?: Function,

  /** The zoom event handler */
  onZoom?: Function,

  /** The zoomend event handler */
  onZoomend?: Function,

  /** The rotatestart event handler */
  onRotatestart?: Function,

  /** The rotate event handler */
  onRotate?: Function,

  /** The rotateend event handler */
  onRotateend?: Function,

  /** The pitchstart event handler */
  onPitchstart?: Function,

  /** The pitch event handler */
  onPitch?: Function,

  /** The pitchend event handler */
  onPitchend?: Function,

  /** The boxzoomstart event handler */
  onBoxzoomstart?: Function,

  /** The boxzoomend event handler */
  onBoxzoomend?: Function,

  /** The boxzoomcancel event handler */
  onBoxzoomcancel?: Function,

  /** The webglcontextlost event handler */
  onWebglcontextlost?: Function,

  /** The webglcontextrestored event handler */
  onWebglcontextrestored?: Function,

  /** The load event handler */
  onLoad?: Function,

  /** The error event handler */
  onError?: Function,

  /** The data event handler */
  onData?: Function,

  /** The styledata event handler */
  onStyledata?: Function,

  /** The sourcedata event handler */
  onSourcedata?: Function,

  /** The dataloading event handler */
  onDataloading?: Function,

  /** The styledataloading event handler */
  onStyledataloading?: Function,

  /** The sourcedataloading event handler */
  onSourcedataloading?: Function
};
