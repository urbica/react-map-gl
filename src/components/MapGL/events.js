const events = [
  'resize',
  'remove',
  'mousedown',
  'mouseup',
  'mouseover',
  'mousemove',
  'click',
  'dblclick',
  'mouseenter',
  'mouseleave',
  'mouseout',
  'contextmenu',
  'touchstart',
  'touchend',
  'touchmove',
  'touchcancel',
  'movestart',
  'move',
  'moveend',
  'dragstart',
  'drag',
  'dragend',
  'zoomstart',
  'zoom',
  'zoomend',
  'rotatestart',
  'rotate',
  'rotateend',
  'pitchstart',
  'pitch',
  'pitchend',
  'boxzoomstart',
  'boxzoomend',
  'boxzoomcancel',
  'webglcontextlost',
  'webglcontextrestored',
  'load',
  'error',
  'data',
  'styledata',
  'sourcedata',
  'dataloading',
  'styledataloading',
  'sourcedataloading'
];

module.exports = events;
