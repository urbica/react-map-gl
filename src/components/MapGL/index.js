// @flow

import { PureComponent, createElement, createRef } from 'react';
import type { ChildrenArray, Element, ElementRef } from 'react';
import type MapboxMap from 'mapbox-gl/src/ui/map';
import type MapboxLngLatBoundsLike from 'mapbox-gl/src/geo/lng_lat_bounds';
import type { AnimationOptions } from 'mapbox-gl/src/ui/camera';
import type { StyleSpecification } from 'mapbox-gl/src/style-spec/types';
import type { MapMouseEvent, MapTouchEvent } from 'mapbox-gl/src/ui/events';

import MapContext from '../MapContext';
import mapboxgl from '../../utils/mapbox-gl';
import events from './events';
import capitalizeFirstLetter from '../../utils/capitalizeFirstLetter';

/* eslint-disable import/no-cycle */
import normalizeChildren from '../../utils/normalizeChildren';

import type { EventProps } from './eventProps';

export type Viewport = {|
  latitude: number,
  longitude: number,
  zoom: number,
  pitch?: number,
  bearing?: number
|};

export const jumpTo = 'jumpTo';
export const easeTo = 'easeTo';
export const flyTo = 'flyTo';

export type ViewportChangeMethod = 'jumpTo' | 'easeTo' | 'flyTo';
export type ViewportChangeEvent = MapMouseEvent | MapTouchEvent;

export type Children = ChildrenArray<Element<any>>;

type Props = EventProps & {
  /** container className */
  className?: string,

  /** container style */
  style?: Object,

  /**
   * The Mapbox style. A string url or a Mapbox GL style object.
   */
  mapStyle: string | StyleSpecification,

  /** MapGL children as Sources, Layers, Controls, and custom Components */
  children?: Children,

  /**
   * Mapbox API access token for mapbox-gl-js.
   * Required when using Mapbox vector tiles/styles.
   */
  accessToken?: string,

  /** The longitude of the center of the map. */
  longitude: number,

  /** The latitude of the center of the map. */
  latitude: number,

  /** The tile zoom level of the map. */
  zoom: number,

  /** Specify the bearing of the viewport */
  bearing?: number,

  /** Specify the pitch of the viewport */
  pitch?: number,

  /** The minimum zoom level of the map (0-22). */
  minZoom?: number,

  /** The maximum zoom level of the map (0-22). */
  maxZoom?: number,

  /**
   * If `true`, the map's position (zoom, center latitude,
   * center longitude, bearing, and pitch) will be synced
   * with the hash fragment of the page's URL. For example,
   * http://path/to/my/page.html#2.59/39.26/53.07/-24.1/60.
   */
  hash?: boolean,

  /**
   * The threshold, measured in degrees, that determines when
   * the map's bearing (rotation) will snap to north. For
   * example, with a  bearingSnap of 7, if the user rotates the
   * map within 7 degrees of north, the map will automatically
   * snap to exact north.
   */
  bearingSnap?: number,

  /**
   * If `false`, the map's pitch (tilt) control with "drag to
   * rotate" interaction will be disabled.
   */
  pitchWithRotate?: boolean,

  /** If `true`, an AttributionControl will be added to the map. */
  attributionControl?: boolean,

  /* A string representing the position of the Mapbox wordmark on the map. */
  logoPosition?: 'top-left' | 'top-right' | 'bottom-left' | 'bottom-right',

  /**
   * If `true`, map creation will fail if the performance of Mapbox
   * GL JS would be dramatically worse than expected (i.e. a software
   * renderer would be used).
   */
  failIfMajorPerformanceCaveat?: boolean,

  /**
   * Mapbox WebGL context creation option.
   * Useful when you want to export the canvas as a PNG.
   */
  preserveDrawingBuffer?: boolean,

  /**
   * If `false`, the map won't attempt to re-request tiles once they
   * expire per their HTTP `cacheControl`/`expires` headers.
   */
  refreshExpiredTiles?: boolean,

  /** If set, the map will be constrained to the given bounds. */
  maxBounds?: MapboxLngLatBoundsLike,

  /** If `true`, the "scroll to zoom" interaction is enabled. */
  scrollZoom?: boolean | Object,

  /** If `true`, the "box zoom" interaction is enabled */
  boxZoom?: boolean,

  /** If `true`, the "drag to rotate" interaction is enabled */
  dragRotate?: boolean,

  /** If `true`, the "drag to pan" interaction is enabled */
  dragPan?: boolean,

  /** If `true`, keyboard shortcuts are enabled */
  keyboard?: boolean,

  /** If `true`, the "double click to zoom" interaction is enabled */
  doubleClickZoom?: boolean,

  /**
   * If `true`, the map will automatically resize
   * when the browser window resizes.
   */
  trackResize?: boolean,

  /**
   * The maximum number of tiles stored in the tile cache for a given
   * source. If omitted, the cache will be dynamically sized based on
   * the current viewport.
   */
  maxTileCacheSize?: number,

  /**
   * If `true`, multiple copies of the world
   * will be rendered, when zoomed out
   */
  renderWorldCopies?: boolean,

  /**
   * If specified, defines a CSS font-family for locally overriding
   * generation of glyphs in the 'CJK Unified Ideographs' and
   * 'Hangul Syllables' ranges. In these ranges, font settings from the
   * map's style will be ignored, except for font-weight keywords
   * (light/regular/medium/bold). The purpose of this option is to avoid
   * bandwidth-intensive glyph server requests.
   * (see https://www.mapbox.com/mapbox-gl-js/example/local-ideographs )
   */
  localIdeographFontFamily?: boolean,

  /**
   * A callback run before the Map makes a request for an external URL.
   * The callback can be used to modify the url, set headers, or set the
   * credentials property for cross-origin requests. Expected to return
   * an object with a  url property and optionally  headers and
   * credentials properties.
   */
  transformRequest?: (
    url: string,
    resourceType: string
  ) => { url: string, headers?: Object, credentials?: string },

  /**
   * If true, Resource Timing API information will be collected for
   * requests made by GeoJSON and Vector Tile web workers (this information
   * is normally inaccessible from the main Javascript thread). Information
   * will be returned in a resourceTiming property of relevant data events.
   */
  collectResourceTiming?: boolean,

  /**
   * `onViewportChange` callback is fired when the user interacted with the
   * map. The object passed to the callback contains viewport properties
   * such as `longitude`, `latitude`, `zoom` etc.
   */
  onViewportChange?: (viewport: Viewport) => void,

  /**
   * Map method that will be called after new viewport props are received.
   */
  viewportChangeMethod?: ViewportChangeMethod,

  /**
   * Options common to map movement methods that involve animation,
   * controlling the duration and easing function of the animation.
   * This options will be passed to the `viewportChangeMethod` call.
   * (see https://docs.mapbox.com/mapbox-gl-js/api/#animationoptions)
   */
  viewportChangeOptions?: AnimationOptions,

  /** The onLoad callback for the map */
  onLoad?: Function,

  /** Map cursor style as CSS value */
  cursorStyle?: string,

  /**
   * Sets a Boolean indicating whether the map will render an outline around
   * each tile and the tile ID. These tile boundaries are useful for debugging.
   * */
  showTileBoundaries?: boolean
};

type State = {
  loaded: boolean
};

class MapGL extends PureComponent<Props, State> {
  _map: MapboxMap;

  _container: { current: null | ElementRef<'div'> };

  _onViewportChange: (event: ViewportChangeEvent) => void;

  static displayName = 'MapGL';

  static defaultProps = {
    children: null,
    className: null,
    style: null,
    mapStyle: 'mapbox://styles/mapbox/light-v8',
    accessToken: null,
    bearing: 0,
    pitch: 0,
    minZoom: 0,
    maxZoom: 22,
    maxBounds: null,
    hash: false,
    bearingSnap: 7,
    pitchWithRotate: true,
    attributionControl: true,
    logoPosition: 'bottom-left',
    failIfMajorPerformanceCaveat: false,
    preserveDrawingBuffer: false,
    refreshExpiredTiles: true,
    boxZoom: true,
    scrollZoom: true,
    dragRotate: true,
    dragPan: true,
    keyboard: true,
    doubleClickZoom: true,
    trackResize: true,
    renderWorldCopies: true,
    maxTileCacheSize: null,
    onViewportChange: null,
    onLoad: null,
    localIdeographFontFamily: null,
    transformRequest: null,
    collectResourceTiming: false,
    cursorStyle: null,
    viewportChangeMethod: jumpTo,
    viewportChangeOptions: null
  };

  constructor(props: Props) {
    super(props);

    if (mapboxgl) {
      mapboxgl.accessToken = props.accessToken;
    }

    this._container = createRef();
  }

  state = {
    loaded: false
  };

  componentDidMount() {
    if (!mapboxgl) {
      return;
    }

    const container = this._container.current;

    const map: MapboxMap = new mapboxgl.Map({
      container,
      style: this.props.mapStyle,
      interactive: !!this.props.onViewportChange,
      center: [this.props.longitude, this.props.latitude],
      zoom: this.props.zoom,
      pitch: this.props.pitch,
      bearing: this.props.bearing,
      minZoom: this.props.minZoom,
      maxZoom: this.props.maxZoom,
      hash: this.props.hash,
      bearingSnap: this.props.bearingSnap,
      pitchWithRotate: this.props.pitchWithRotate,
      attributionControl: this.props.attributionControl,
      logoPosition: this.props.logoPosition,
      failIfMajorPerformanceCaveat: this.props.failIfMajorPerformanceCaveat,
      preserveDrawingBuffer: this.props.preserveDrawingBuffer,
      refreshExpiredTiles: this.props.refreshExpiredTiles,
      maxBounds: this.props.maxBounds,
      scrollZoom: this.props.scrollZoom,
      boxZoom: this.props.boxZoom,
      dragRotate: this.props.dragRotate,
      dragPan: this.props.dragPan,
      keyboard: this.props.keyboard,
      doubleClickZoom: this.props.doubleClickZoom,
      trackResize: this.props.trackResize,
      renderWorldCopies: this.props.renderWorldCopies,
      maxTileCacheSize: this.props.maxTileCacheSize,
      localIdeographFontFamily: this.props.localIdeographFontFamily,
      transformRequest: this.props.transformRequest,
      collectResourceTiming: this.props.collectResourceTiming
    });

    this._map = map;

    map.on('styledata', () => this.setState({ loaded: true }));

    map.once('load', () => {
      this.setState({ loaded: true }, this.props.onLoad);
    });

    if (this.props.onViewportChange) {
      map.on('dragend', this._onViewportChange);
      map.on('zoomend', this._onViewportChange);
      map.on('rotateend', this._onViewportChange);
      map.on('pitchend', this._onViewportChange);
      map.on('boxzoomend', this._onViewportChange);
    }

    events.forEach((event) => {
      const propName = `on${capitalizeFirstLetter(event)}`;
      if (this.props[propName]) {
        map.on(event, this.props[propName]);
      }
    });

    if (this.props.cursorStyle) {
      map.getCanvas().style.cursor = this.props.cursorStyle;
    }

    if (this.props.showTileBoundaries) {
      this._map.showTileBoundaries = this.props.showTileBoundaries;
    }
  }

  componentDidUpdate(prevProps: Props) {
    this._updateMapViewport(prevProps, this.props);
    this._updateMapStyle(prevProps, this.props);

    if (!prevProps.cursorStyle !== this.props.cursorStyle) {
      this._map.getCanvas().style.cursor = this.props.cursorStyle;
    }

    if (prevProps.showTileBoundaries !== this.props.showTileBoundaries) {
      this._map.showTileBoundaries = !!this.props.showTileBoundaries;
    }
  }

  componentWillUnmount() {
    if (this._map) {
      this._map.remove();
    }
  }

  // External apps can access map this way
  getMap() {
    return this._map;
  }

  /**
   * Update Map style from newProps
   *
   * @private
   * @param {Props} prevProps
   * @param {Props} newProps
   */
  _updateMapStyle(prevProps: Props, newProps: Props): void {
    const newMapStyle = newProps.mapStyle;
    const prevMapStyle = prevProps.mapStyle;

    if (newMapStyle !== prevMapStyle) {
      this.setState({ loaded: false }, () => this._map.setStyle(newMapStyle));
    }
  }

  /**
   * Update Map viewport from newProps
   *
   * @private
   * @param {Props} prevProps
   * @param {Props} newProps
   */
  _updateMapViewport(prevProps: Props, newProps: Props): void {
    const map: MapboxMap = this._map;
    const center = map.getCenter();

    const viewportChanged =
      (newProps.latitude !== prevProps.latitude &&
        newProps.latitude !== center.lat) ||
      (newProps.longitude !== prevProps.longitude &&
        newProps.longitude !== center.lng) ||
      (newProps.zoom !== prevProps.zoom && newProps.zoom !== map.getZoom()) ||
      (newProps.pitch !== prevProps.pitch &&
        newProps.pitch !== map.getPitch()) ||
      (newProps.bearing !== prevProps.bearing &&
        newProps.bearing !== map.getBearing());

    if (!viewportChanged) {
      return;
    }

    const viewport = {
      center: [newProps.longitude, newProps.latitude],
      zoom: newProps.zoom,
      pitch: newProps.pitch,
      bearing: newProps.bearing
    };

    const { viewportChangeMethod, viewportChangeOptions } = this.props;
    const options = { ...viewportChangeOptions, ...viewport };

    switch (viewportChangeMethod) {
      case flyTo:
        map.flyTo(options);
        break;
      case jumpTo:
        map.jumpTo(options);
        break;
      case easeTo:
        map.easeTo(options);
        break;
      default:
        throw new Error('Unknown viewport change method');
    }
  }

  /**
   * fires `onViewportChange` callback when the user interacted with the map.
   *
   * @private
   * @param {ViewportChangeEvent} event
   */
  _onViewportChange = (event: ViewportChangeEvent): void => {
    if (!event.originalEvent) {
      return;
    }

    const map = this._map;

    const { lng, lat } = map.getCenter();
    const zoom = map.getZoom();
    const pitch = map.getPitch();
    const bearing = map.getBearing();

    const viewport: Viewport = {
      latitude: lat,
      longitude: lng,
      zoom,
      pitch,
      bearing
    };

    // $FlowFixMe
    this.props.onViewportChange(viewport);
  };

  render() {
    const { loaded } = this.state;
    const { className, style } = this.props;

    const children = this.props.children
      ? normalizeChildren(this.props.children)
      : null;

    return createElement(
      MapContext.Provider,
      { value: this._map },
      createElement(
        'div',
        {
          ref: this._container,
          style,
          className
        },
        loaded && children
      )
    );
  }
}

export default MapGL;
