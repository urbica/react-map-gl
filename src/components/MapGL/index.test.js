/* eslint-disable no-console */

import React from 'react';
import { mount } from 'enzyme';

import MapGL from './index';
import { CustomLayer, Layer, Source } from '../../index';

test('render', () => {
  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0} cursorStyle="pointer" />
  );

  expect(wrapper.exists()).toBe(true);
  const map = wrapper.instance().getMap();
  expect(map).toBeTruthy();

  wrapper.unmount();
  expect(wrapper.exists()).toBe(false);
  expect(map.remove).toHaveBeenCalled();
});

test('onLoad', () => {
  const onLoad = jest.fn();
  mount(<MapGL latitude={0} longitude={0} zoom={0} onLoad={onLoad} />);
  expect(onLoad).toHaveBeenCalled();
  // expect(onLoad).toHaveBeenCalledTimes(1);
});

test('onViewportChange', () => {
  const onViewportChange = jest.fn();

  mount(
    <MapGL
      latitude={0}
      longitude={0}
      zoom={0}
      onViewportChange={onViewportChange}
    />
  );

  expect(onViewportChange).toHaveBeenCalled();
  expect(onViewportChange).toHaveBeenCalledTimes(5);
});

test('viewport update', () => {
  const wrapper = mount(<MapGL latitude={0} longitude={0} zoom={0} />);

  wrapper.setProps({ latitude: 1 });
  expect(wrapper.props().latitude).toBe(1);

  wrapper.setProps({ longitude: 2 });
  expect(wrapper.props().longitude).toBe(2);

  wrapper.setProps({ zoom: 3 });
  expect(wrapper.props().zoom).toBe(3);

  wrapper.setProps({ pitch: 4 });
  expect(wrapper.props().pitch).toBe(4);

  wrapper.setProps({ bearing: 5 });
  expect(wrapper.props().bearing).toBe(5);
});

test('viewportChangeMethod update', () => {
  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0} viewportChangeMethod="jumpTo" />
  );

  wrapper.setProps({ zoom: 1, viewportChangeMethod: 'flyTo' });
  expect(wrapper.props().viewportChangeMethod).toBe('flyTo');

  wrapper.setProps({ zoom: 2, viewportChangeMethod: 'easeTo' });
  expect(wrapper.props().viewportChangeMethod).toBe('easeTo');
});

test('viewportChangeMethod throw', () => {
  console.error = jest.fn();

  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0} viewportChangeMethod="jumpTo" />
  );

  expect(() => {
    wrapper.setProps({ zoom: 3, viewportChangeMethod: 'invalid' });
  }).toThrow();

  expect(console.error).toHaveBeenCalled();
});

test('mapStyle update', () => {
  const wrapper = mount(
    <MapGL
      mapStyle="mapbox://styles/mapbox/light-v9"
      latitude={0}
      longitude={0}
      zoom={0}
    />
  );

  wrapper.setProps({ mapStyle: 'mapbox://styles/mapbox/dark-v9' });
  expect(wrapper.props().mapStyle).toBe('mapbox://styles/mapbox/dark-v9');
});

test('multiple layers', () => {
  const data = { type: 'FeatureCollection', features: [] };

  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0}>
      <Source id="test" type="geojson" data={data} />
      <Layer id="test1" type="circle" source="test" />
      <Layer id="test2" type="circle" source="test" />
    </MapGL>
  );

  expect(wrapper.find('Layer').exists()).toBe(true);
  expect(wrapper.find('Layer')).toHaveLength(2);

  wrapper.unmount();
  expect(wrapper.find('Layer').exists()).toBe(false);
});

test('children order', () => {
  const data = { type: 'FeatureCollection', features: [] };

  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0}>
      <Source id="test1" type="geojson" data={data} />
      <Layer id="test1" type="circle" source="test1" />
      <Source id="test2" type="geojson" data={data} />
      <Layer id="test2" type="circle" source="test2" />
      <Source id="test3" type="geojson" data={data} />
      <Layer id="test3" type="circle" source="test3" />
    </MapGL>
  );

  const children = wrapper.find('div');
  expect(children.childAt(0).type()).toBe(Source);
  expect(children.childAt(1).type()).toBe(Layer);
  expect(children.childAt(2).type()).toBe(Source);
  expect(children.childAt(3).type()).toBe(Layer);
  expect(children.childAt(4).type()).toBe(Source);
  expect(children.childAt(5).type()).toBe(Layer);
});

test('layers ordering', () => {
  const data = { type: 'FeatureCollection', features: [] };

  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0}>
      <Source id="test1" type="geojson" data={data} />
      <Layer id="test1" type="circle" source="test1" />
      <Source id="test2" type="geojson" data={data} />
      <Layer id="test2" type="circle" source="test2" />
      <Source id="test3" type="geojson" data={data} />
      <Layer id="test3" type="circle" source="test3" />
    </MapGL>
  );

  const layersWrapper = wrapper.find('Layer');
  expect(layersWrapper.find({ id: 'test1' }).props().before).toBe('test2');
  expect(layersWrapper.find({ id: 'test2' }).props().before).toBe('test3');
  expect(layersWrapper.find({ id: 'test3' }).props().before).toBe(undefined);
});

test('normalizeChildren', () => {
  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0}>
      <Layer id="layer1" />
      <Layer id="layer2" />
      <Layer id="layer3" />
      {null}
      <Source id="source1" type="vector">
        <Layer id="layer4" />
        {null}
        <Layer id="layer5" />
      </Source>
      {null}
      <CustomLayer layer={{ id: 'layer6' }} />
      <React.Fragment>
        <Layer id="layer7" />
        {null}
        <Layer id="layer8" />
      </React.Fragment>
      <Layer id="layer0" before="layer1" />
      {null}
      <Source id="source2" type="vector">
        <Layer id="layer9" />
        <Layer id="layer10" />
      </Source>
    </MapGL>
  );

  const layersWrapper = wrapper.find('Layer');
  expect(layersWrapper.find({ id: 'layer1' }).props().before).toBe('layer2');
  expect(layersWrapper.find({ id: 'layer2' }).props().before).toBe('layer3');
  expect(layersWrapper.find({ id: 'layer3' }).props().before).toBe('layer4');
  expect(layersWrapper.find({ id: 'layer4' }).props().before).toBe('layer5');
  expect(layersWrapper.find({ id: 'layer5' }).props().before).toBe('layer6');
  expect(
    wrapper
      .find('CustomLayer')
      .find({ layer: { id: 'layer6' } })
      .props().before
  ).toBe('layer7');
  expect(layersWrapper.find({ id: 'layer7' }).props().before).toBe('layer8');
  expect(layersWrapper.find({ id: 'layer8' }).props().before).toBe('layer9');
  expect(layersWrapper.find({ id: 'layer9' }).props().before).toBe('layer10');
  expect(layersWrapper.find({ id: 'layer10' }).props().before).toBe(undefined);
});

test('multiple sources', () => {
  const data = { type: 'FeatureCollection', features: [] };

  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0}>
      <Source
        id="test1"
        type="vector"
        url="mapbox://mapbox.mapbox-terrain-v2"
      />
      <Source id="test2" type="geojson" data={data} />
    </MapGL>
  );

  expect(wrapper.find('Source').exists()).toBe(true);
  expect(wrapper.find('Source')).toHaveLength(2);

  wrapper.unmount();
  expect(wrapper.find('Source').exists()).toBe(false);
});

test('circular add/remove layers', () => {
  const data = { type: 'FeatureCollection', features: [] };

  class Wrapper extends React.PureComponent {
    state = {
      test1: true,
      test2: false
    };

    render() {
      return (
        <MapGL latitude={0} longitude={0} zoom={0}>
          {this.state.test1 && (
            <React.Fragment>
              <Source id="test1" type="geojson" data={data} />
              <Layer id="test1" type="circle" source="test1" />
            </React.Fragment>
          )}
          {this.state.test2 && (
            <React.Fragment>
              <Source id="test2" type="geojson" data={data} />
              <Layer id="test2" type="circle" source="test2" />
            </React.Fragment>
          )}
        </MapGL>
      );
    }
  }

  const wrapper = mount(<Wrapper />);

  expect(() => {
    wrapper.setState({ test2: true });
    wrapper.setState({ test1: false });
    wrapper.setState({ test1: true });
    wrapper.setState({ test2: false });
  }).not.toThrow();
});

test('renders with tileBoundaries', () => {
  const wrapper = mount(
    <MapGL latitude={0} longitude={0} zoom={0} showTileBoundaries />
  );

  const map = wrapper.instance().getMap();
  expect(map.showTileBoundaries).toBe(true);

  wrapper.setProps({ showTileBoundaries: false });
  expect(map.showTileBoundaries).toBe(false);
});

test('do not call onViewportChange if originalEvent is not present', () => {
  /* eslint-disable global-require */
  const mapboxgl = require('../../__mocks__/mapbox-gl');
  mapboxgl.Map.prototype.on = function on(_, listener, fn) {
    const handler = typeof listener === 'function' ? listener : fn;
    handler({ target: this, originalEvent: false, point: { x: 0, y: 0 } });
  };
  jest.setMock('mapbox-gl', mapboxgl);

  const onViewportChange = jest.fn();

  mount(
    <MapGL
      latitude={0}
      longitude={0}
      zoom={0}
      onViewportChange={onViewportChange}
    />
  );

  expect(onViewportChange).not.toHaveBeenCalled();
});

test('supports array event handlers', () => {
  const onClick = jest.fn();

  const wrapper = mount(
    <MapGL
      latitude={0}
      longitude={0}
      zoom={0}
      onClick={['national_park', onClick]}
    />
  );

  expect(wrapper.exists()).toBe(true);
});

test('renders without mapbox-gl', () => {
  jest.resetModules();
  jest.doMock('mapbox-gl', () => null);

  /* eslint-disable no-shadow, global-require */
  const MapGL = require('.').default;

  const wrapper = mount(<MapGL latitude={0} longitude={0} zoom={0} />);

  expect(wrapper.exists()).toBe(true);
  const map = wrapper.instance().getMap();
  expect(map).toBeFalsy();
});
