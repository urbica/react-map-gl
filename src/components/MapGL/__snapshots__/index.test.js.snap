// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`normalizeChildren 1`] = `
<MapGL
  accessToken={null}
  attributionControl={true}
  bearing={0}
  bearingSnap={7}
  boxZoom={true}
  className={null}
  collectResourceTiming={false}
  cursorStyle={null}
  doubleClickZoom={true}
  dragPan={true}
  dragRotate={true}
  failIfMajorPerformanceCaveat={false}
  hash={false}
  keyboard={true}
  latitude={0}
  localIdeographFontFamily={null}
  logoPosition="bottom-left"
  longitude={0}
  mapStyle="mapbox://styles/mapbox/light-v8"
  maxBounds={null}
  maxTileCacheSize={null}
  maxZoom={22}
  minZoom={0}
  onLoad={null}
  onViewportChange={null}
  pitch={0}
  pitchWithRotate={true}
  preserveDrawingBuffer={false}
  refreshExpiredTiles={true}
  renderWorldCopies={true}
  scrollZoom={true}
  style={null}
  trackResize={true}
  transformRequest={null}
  viewportChangeMethod="jumpTo"
  viewportChangeOptions={null}
  zoom={0}
>
  <div
    className={null}
    style={null}
  >
    <Layer
      before="layer2"
      id="layer1"
      key=".$.0"
      radius={0}
    />
    <Layer
      before="layer3"
      id="layer2"
      key=".$.1"
      radius={0}
    />
    <Layer
      before="layer4"
      id="layer3"
      key=".$.2"
      radius={0}
    />
    <Source
      id="source1"
      key=".$.3"
      type="vector"
    >
      <Layer
        before="layer5"
        id="layer4"
        key=".0"
        radius={0}
      />
      <Layer
        before="layer6"
        id="layer5"
        key=".1"
        radius={0}
      />
    </Source>
    <CustomLayer
      before="layer7"
      key=".$.4"
      layer={
        Object {
          "id": "layer6",
        }
      }
    />
    <Layer
      before="layer8"
      id="layer7"
      key=".0"
      radius={0}
    />
    <Layer
      before="layer9"
      id="layer8"
      key=".1"
      radius={0}
    />
    <Layer
      before="layer1"
      id="layer0"
      key=".$.6"
      radius={0}
    />
    <Source
      id="source2"
      key=".$.7"
      type="vector"
    >
      <Layer
        before="layer10"
        id="layer9"
        key=".0"
        radius={0}
      />
      <Layer
        id="layer10"
        key=".1"
        radius={0}
      />
    </Source>
  </div>
</MapGL>
`;
